### PR TITLE
[MM-39510] Don't provide channel, team and post ID by default in call context

### PR DESCRIPTION
**Pull request by hanzei**
#### Summary
As a foundation of the bindings caching work, channel, team and post ID are not longer provided by default in the call context. To get the ID, an app most specific a new `ExpandLevel` `ExpandID`, which will request as part of the context. It will be place inside of `channel` etc. not `channel_id`.

The PR also moved some fields from `Context` to `ExpandedContext` that are set by the proxy and hence belong in `ExpandedContext`.

Lastly, the ordering of fields got updated. The order is now: app static information, acting user info, channel info, team info, post info, call semantics like mentions, oauth2 info. This is is used consistently now.

Heads that that the PR targets #291 to avoid merge conflicts.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-39510
https://mattermost.atlassian.net/browse/MM-39935

### Breaking Changes
- Removes `ChannelID`, `TeamID`, `PostID`, `RootPostID` and `PostID` from outgoing requests

The following apps need to get updated:
- [x] Examples
- [x] Test app: https://github.com/mattermost/mattermost-app-test/pull/13
- [x] ServiceNow: https://github.com/mattermost/mattermost-app-servicenow/pull/36
- [x] Zendesk: https://github.com/mattermost/mattermost-app-zendesk/pull/90
- [ ] gCal

### CODE REVIEW

**Comment by hanzei** [comment]
@levb @mickmister Gentle reminder to review this PR

File: apps/appclient/mattermost_client.go [hunk 1]
```diff
@@ -25,7 +25,7 @@ func AsBot(cc apps.Context) *Client {
 }
 
 func AsActingUser(cc apps.Context) *Client {
-	return as(cc.ActingUserID, cc.ActingUserAccessToken, cc)
+	return as(cc.ActingUser.Id, cc.ActingUserAccessToken, cc)
```
Comment (levb) [inline apps/appclient/mattermost_client.go:28]: What is the rationale for the change here? We'd need to check for `ActingUser != nil`, no?

File: apps/expand.go [hunk 2]
```diff
@@ -49,34 +54,30 @@ type Expand struct {
...
 
+	// Post, RootPost: all for model.Post, summary for Id, Type, UserId,
+	// ChannelId, RootId, Message.
+	Post     ExpandLevel `json:"post,omitempty"`
+	RootPost ExpandLevel `json:"root_post,omitempty"`
+
 	// User: all for model.User, summary for BotDescription, DeleteAt, Email,
 	// FirstName, Id, IsBot, LastName, Locale, Nickname, Roles, Timezone,
-	// Username.
+	// Username. Default is summary.
```
Comment (levb) [inline apps/expand.go:71]: 👍

**Review by levb — COMMENTED** [review]
I am a little confused about `ActingUserID` - if it is there, why is it not used; or why is it still there?

Otherwise, LGTM

**Comment by hanzei** [comment]
@levb @mickmister @catalintomai I would appreciate a review on this PR as it's another breaking change that impacts the other apps.

**Comment by hanzei** [comment]
Sorry, had to force push to fix the history.

File: apps/context.go [hunk 3]
```diff
@@ -14,30 +14,24 @@ import (
...
-	// subscription.
-	Subject Subject `json:"subject,omitempty"`
-
-	// BotUserID of the App.
-	BotUserID string `json:"bot_user_id"`
-
 	// ActingUserID is primarily (or exclusively?) for calls originating from
 	// user submissions.
+	//
+	// UserID is not send down to Apps.
```
Comment (catalintomai) [inline apps/context.go:23]: UserID -> ActingUserID

Comment (hanzei) [reply apps/context.go]: Fixed

**Review by catalintomai — APPROVED** [review]

**Comment by hanzei** [comment]
I'm going to merge this PR to simply further development and testing

**Comment by hanzei** [comment]
Documented in https://developers.mattermost.com/integrate/apps/structure/call-metadata/
